### PR TITLE
Hierarchical concepts and CodeCaretValueRule support

### PR DESCRIPTION
**Pull request by julianxcarter**
Addresses [CIMPL-744](https://standardhealthrecord.atlassian.net/browse/CIMPL-744), adding support for hierarchical concepts and concepts with `designation` or `property` values assigned. These newly supported concepts are also indented by the `SimplifyRulePathContextsOptimizer`.

### CODE REVIEW

File: src/extractor/CaretValueRuleExtractor.ts [hunk 1]
```diff
@@ -231,6 +240,29 @@ export class CaretValueRuleExtractor {
...
+      if (isFSHValueEmpty(rule.value)) {
+        logger.error(
+          `Value in CodeSytem ${
+            codeSystem.name ?? codeSystem.id
+          } at concept ${conceptHeirarchy.join('.')} for element ${
+            rule.caretPath
+          } is empty. No caret value rule will be created.`
+        );
+        delete caretValueRules[index];
+      }
```
Comment (mint-thompson) [inline src/extractor/CaretValueRuleExtractor.ts:261]: Could you add a test case for an empty value in a concept's caret rule? I don't think any of the current test cases will go into this block.

**Review by mint-thompson — CHANGES_REQUESTED** [review]
Hooray for hierarchical concepts! I have a few small changes to request.

File: src/extractor/CaretValueRuleExtractor.ts [hunk 2]
```diff
@@ -231,6 +240,29 @@ export class CaretValueRuleExtractor {
     return caretValueRules;
   }
 
+  static processConcept(
+    input: any,
+    conceptHeirarchy: string[],
```
[the remark this answers is not in the document]
Comment (julianxcarter) [reply src/extractor/CaretValueRuleExtractor.ts]: Whoops! Thanks for catching these, this should be addressed in commit `6fe53f2`.

File: src/extractor/CaretValueRuleExtractor.ts [hunk 3]
```diff
@@ -218,6 +222,11 @@ export class CaretValueRuleExtractor {
         const caretValueRule = new ExportableCaretValueRule('');
         caretValueRule.caretPath = key;
         caretValueRule.value = getFSHValue(key, flatVS, resourceType, fisher);
+        if (resourceType === 'Concept') {
+          caretValueRule.isCodeCaretRule = true;
+          caretValueRules.push(caretValueRule);
+          return;
+        }
```
[the remark this answers is not in the document]
Comment (julianxcarter) [reply src/extractor/CaretValueRuleExtractor.ts]: In commit `a1b231b` I opted to just get rid of the `processConcept` function and handle everything in the `processResource` function.

File: src/extractor/CaretValueRuleExtractor.ts (hunk 1 above)
Comment (julianxcarter) [reply src/extractor/CaretValueRuleExtractor.ts]: Added in commit `a1b231b`, along with some other unit tests.

**Review by ngfreiter — CHANGES_REQUESTED** [review]
I think something seems weird with `package-lock.json`. Why is it changing when this doesn't update `package.json`. And shouldn't `package.json` now point to `2.0.0-beta.3`? Or else I think the new code is referencing properties that are not in `2.0.0-beta.1`

File: src/extractor/CaretValueRuleExtractor.ts [hunk 4]
```diff
@@ -199,16 +199,19 @@ export class CaretValueRuleExtractor {
...
     const flatVS = getPathValuePairs(input);
     Object.keys(flatVS)
       .filter(
         key =>
           !RESOURCE_IGNORED_PROPERTIES[resourceType].some(
-            property => key === property || new RegExp(`${property}(\\[\\d+\\])?\\.`).test(key)
+            property =>
+              key === property ||
+              new RegExp(`${property}(\\[\\d+\\])?\\.`).test(key) ||
+              key.startsWith(`${property}[`)
```
Comment (ngfreiter) [inline src/extractor/CaretValueRuleExtractor.ts:214]: This seems really similar to the line above, I'm wondering if maybe something like: 
```
new RegExp(`${property}(\\[\\d+\\]|\\.)`).test(key)
```
could capture all of the cases we need.

File: src/extractor/CaretValueRuleExtractor.ts [hunk 5]
```diff
@@ -199,16 +199,19 @@ export class CaretValueRuleExtractor {
   static processResource(
     input: any,
     fisher: utils.Fishable,
-    resourceType: 'ValueSet' | 'CodeSystem',
-    config: fshtypes.Configuration
+    resourceType: 'ValueSet' | 'CodeSystem' | 'Concept',
```
Comment (ngfreiter) [inline src/extractor/CaretValueRuleExtractor.ts:202]: I don't think we want to do this, the function is `processResource`, and the param is `resourceType`, so it feels off to accept a `Concept` here, since Concepts are not a type of FHIR resource.

Comment (ngfreiter) [reply src/extractor/CaretValueRuleExtractor.ts:202]: And you end up having to do a solid amount of special casing for `Concept` anyway, so I think it's probably best to just make this its own thing.

File: src/processor/CodeSystemProcessor.ts [hunk 6]
```diff
@@ -31,11 +32,40 @@ export class CodeSystemProcessor {
...
+  ): (ExportableCaretValueRule | ExportableConceptRule)[] {
+    const conceptRules: (ExportableConceptRule | ExportableCaretValueRule)[] = [];
+    const newConceptRule = new ExportableConceptRule(
+      concept.code,
+      concept.display,
+      concept.definition
+    );
+    newConceptRule.hierarchy = hierarchy ?? [];
+    concept._parentId = codeSystem.name ?? codeSystem.id;
+    concept._hierarchy = [...newConceptRule.hierarchy, concept.code];
```
Comment (ngfreiter) [inline src/processor/CodeSystemProcessor.ts:54]: I think the other advantage of making a new `processConcept` function on the `CaretValueRuleExtractor` is that then you can just (I think) pass hierarchy and the parent ID as params, which makes things a little bit more readable.

**Review by ngfreiter — CHANGES_REQUESTED** [review]

File: package.json [hunk 7]
```diff
@@ -80,4 +80,4 @@
     "winston": "^3.3.3",
     "yaml": "^1.10.0"
   }
-}
+}
```
Comment (ngfreiter) [inline package.json:83]: Prettier doesn't like this for me, and puts this newline back when I save.

File: package.json [hunk 8]
```diff
@@ -68,7 +68,7 @@
     "fhir": "^4.8.3",
     "flat": "^5.0.2",
     "fs-extra": "^9.0.1",
-    "fsh-sushi": "^2.0.0-beta.1",
+    "fsh-sushi": "^2.0.0-beta.3",
```
Comment (ngfreiter) [inline package.json:71]: I think the package-lock.json is still a little haywire. If I check out this branch and `npm install`, I get a `package-lock.json` with changes. When you run `npm install`, does your lock file change as well? If so I think we want to go ahead and commit those changes.

File: src/extractor/CaretValueRuleExtractor.ts [hunk 9]
```diff
@@ -200,7 +200,7 @@ export class CaretValueRuleExtractor {
     input: any,
     fisher: utils.Fishable,
     resourceType: 'ValueSet' | 'CodeSystem',
-    config: fshtypes.Configuration
+    config?: fshtypes.Configuration
```
Comment (ngfreiter) [inline src/extractor/CaretValueRuleExtractor.ts:203]: This can be reverted now right?

File: src/extractor/CaretValueRuleExtractor.ts [hunk 10]
```diff
@@ -231,6 +231,42 @@ export class CaretValueRuleExtractor {
...
   }
 
+  static processConcept(
+    input: any,
+    conceptHierarchy: string[],
+    codeSystemName: string,
+    fisher: utils.Fishable
+  ): ExportableCaretValueRule[] {
+    const caretValueRules: ExportableCaretValueRule[] = [];
+    const flatVS = getPathValuePairs(input);
```
Comment (ngfreiter) [inline src/extractor/CaretValueRuleExtractor.ts:241]: Probably want to rename this variable.

**Review by ngfreiter — CHANGES_REQUESTED** [review]

File: src/extractor/CaretValueRuleExtractor.ts (hunk 9 above)
Comment (julianxcarter) [reply src/extractor/CaretValueRuleExtractor.ts:203]: It can!

File: package.json (hunk 8 above)
Comment (julianxcarter) [reply package.json:71]: I don't get any changes. I think the issue is that I'm running npm version 7 which creates a `package-lock.json` in a format that's ever so slightly different (which can be viewed as the `lockfileVersion` property at the top of the file). This new version is apparently backwards compatible, but I can downgrade my npm version and run `npm install` again for consistency's sake.

File: src/processor/CodeSystemProcessor.ts [hunk 11]
```diff
@@ -31,11 +32,43 @@ export class CodeSystemProcessor {
...
     );
     input.concept?.forEach((concept: any) => {
-      newRules.push(new ExportableConceptRule(concept.code, concept.display, concept.definition));
+      newRules.push(...this.extractConceptRules(concept, input.name || input.id, fisher));
     });
     target.rules = compact(newRules);
   }
 
+  static extractConceptRules(
+    concept: any,
```
Comment (ngfreiter) [inline src/processor/CodeSystemProcessor.ts:41]: I think we could use the `fhirtypes.CodeSystemConcept` type here that we've nicely defined in SUSHI. Everything should still work fine, but we just get some of the type checking advantages in the future.

File: src/extractor/CaretValueRuleExtractor.ts [hunk 12]
```diff
@@ -231,6 +231,42 @@ export class CaretValueRuleExtractor {
...
+      .filter(
+        key =>
+          !CONCEPT_IGNORED_PROPERTIES.some(
+            property => key === property || new RegExp(`${property}(\\[\\d+\\])?\\.`).test(key)
+          )
+      )
+      .forEach(key => {
+        const caretValueRule = new ExportableCaretValueRule('');
+        caretValueRule.caretPath = key;
+        caretValueRule.value = getFSHValue(key, flatConcept, 'Concept', fisher);
```
Comment (ngfreiter) [inline src/extractor/CaretValueRuleExtractor.ts:252]: I think here you could up front change this to 
```
        caretValueRule.value = getFSHValue(`concept.${key}`, flatConcept, 'CodeSystem', fisher);
```
And then you would not need to change the `getFSHValue` function at all I think?

File: test/optimizer/plugins/fixtures/unsupported-codesystem.json [hunk 13]
```diff
@@ -27,4 +25,4 @@
       }
     }
   ]
-}
+}
```
Comment (ngfreiter) [inline test/optimizer/plugins/fixtures/unsupported-codesystem.json:28]: Prettier is mad again. Prettier needs a constant diet of newlines.

**Review by ngfreiter — COMMENTED** [review]
Looking good, just some final small comments!

**Review by mint-thompson — COMMENTED** [review]
Nothing to report beyond what @ngfreiter already wrote, so I'll re-review once you handle those findings.

File: src/extractor/CaretValueRuleExtractor.ts (hunk 12 above)
Comment (julianxcarter) [reply src/extractor/CaretValueRuleExtractor.ts:252]: This is a clever idea, but doing it this way makes getting the value within the `flatConcept` [on the first line of the `getFSHValue` function](https://github.com/FHIR/GoFSH/blob/14582d5c0ace45370328cd50f15e6edaddee9ae4/src/utils/element.ts#L56) impossible with the key parameter. I could do some clever string manipulation in that function, or change the keys within the `flatConcept` object to also have `concept.` prefixes, but I thought that both of these options felt a bit less intuitive than what I'm currently doing in the `getFSHValue` function. If you prefer either of these approaches then I can make the change!

Comment (ngfreiter) [reply src/extractor/CaretValueRuleExtractor.ts:252]: Ah good point. How it is now is fine!

**Review by ngfreiter — APPROVED** [review]

**Review by mint-thompson — APPROVED** [review]
